shell/platform/linux/fl_binary_messenger.cc

```diff
@@ -141,19 +143,6 @@ static void platform_message_handler_free(gpointer data) {
   g_free(self);
 }
 
-static void engine_weak_notify_cb(gpointer user_data,
-                                  GObject* where_the_object_was) {
-  FlBinaryMessengerImpl* self = FL_BINARY_MESSENGER_IMPL(user_data);
-  self->engine = nullptr;
-
-  // Disconnect any handlers.
-  // Take the reference in case a handler tries to modify this table.
-  g_autoptr(GHashTable) handlers = self->platform_message_handlers;
-  self->platform_message_handlers = g_hash_table_new_full(
-      g_str_hash, g_str_equal, g_free, platform_message_handler_free);
-  g_hash_table_remove_all(handlers);
-}
-
```

Comment:
The removal of this code may be a problem - this being used by plugins to dispose of themselves when the engine stops. We can't detect this using `GWeakRef`, so a possible solution might be for the `FlEngine.dispose` to call a private method on the `FlBinaryMessenger` to trigger this. In practice this might not be an issue, as the disposal of the engine might be the last reference to the messenger and cause it to dispose immediately. It's probably safer to keep the existing behaviour though and it will help if the plugin has taken a reference to the messenger.

See https://github.com/flutter/plugins/blob/main/packages/file_selector/file_selector_linux/linux/file_selector_plugin.cc#L236 for the plugin beahviour - note the only reference to the plugin is held by the binary messenger so the plugin remains alive until either the engine or messenger objects are disposed.

@stuartmorgan may have an opinion on this as he did this plugin architecture with me.

Reply:
When I looked at this; here is why I believed it was okay: 

So this code clears out the hash table when the engine is deleted.  That hash table is also cleared out in `fl_binary_messenger_impl_dispose`.  So removing this code affects the logic only in the time period between the engine being disposed and the binary messenger being disposed.  For example, an async response that will get called after the engine is deleted.  Now if the binary messenger is just a proxy to the engine where stuff really happens, if we have a weak references to the engine and we are guarding any communication to the engine with it, what is the harm with handlers existing for as long as the binary messenger exists if the calls get dropped anyway?  Plus, if the behavior was so important, why is there no test asserting it?

That was my thought process.  If we think this behavior is important to maintain it should be noted why in the code, since it isn't clear by the code alone, and we should have a test asserting it.  From what I see it seems like a slight optimization for memory usage with no logical significance (that I could see).  I think technically it could have worse memory performance if platform message's responses aren't invoked after the engine is killed.  I'm not sure if that was the concern?

Regardless, if we decide this is something that is important to keep I don't think we should add it back without a test asserting it which I might need your help with.

Reply:
> So this code clears out the hash table when the engine is deleted. That hash table is also cleared out in `fl_binary_messenger_impl_dispose`. So removing this code affects the logic only in the time period between the engine being disposed and the binary messenger being disposed.

It doesn't just change the logic during that time, it can change the timing of it happening at all, because there are references involved.

> Now if the binary messenger is just a proxy to the engine where stuff really happens, if we have a weak references to the engine and we are guarding any communication to the engine with it, what is the harm with handlers existing for as long as the binary messenger exists if the calls get dropped anyway?

An unbroken retain cycle between the messenger and the plugins would be harmful because it would leak plugins forever. If we aren't going to have an explicit point where we break them (the code you are removing) we need to move the ownership (e.g., to have the ownership be in the engine, rather than the messenger, as with your engine/messenger split in the iOS code.)

Reply:
Okay, the logic was brought back and a test was added to assert it.